File: proposed-projects/round-trip-experiment/reverse/20260909T022216Z_19e0fe30/20260910T055411Z_259d4a57/ported_implementation/gbnf/grammar_graph/get_serialized_rule_key.py
import json

from .grammar_graph_types import UnresolvedRule
from .type_guards import is_rule_char, is_rule_char_exclude, is_rule_end, is_rule_ref


class KEY_TRANSLATION:
    END = 0
    CHAR = 1
    CHAR_EXCLUDE = 2
    REF = 3


def get_serialized_rule_key(rule: UnresolvedRule) -> str:
    if is_rule_end(rule):
        return f"{KEY_TRANSLATION.END}"

    if is_rule_char(rule):
        return f"{KEY_TRANSLATION.CHAR}-{json.dumps(rule.value)}"

    if is_rule_char_exclude(rule):
        return f"{KEY_TRANSLATION.CHAR_EXCLUDE}-{json.dumps(rule.value)}"

    if is_rule_ref(rule):
        return f"{KEY_TRANSLATION.REF}-{rule.value}"

    raise ValueError(f"Unknown rule type: {rule}")

File: packages/gbnf-experiment/data/20260909T142951Z_4ff7cdec/ported_implementation/src/grammar-graph/get-serialized-rule-key.ts
import { RuleType } from './grammar-graph-types.ts';
import type { UnresolvedRule } from './grammar-graph-types.ts';
import {
  isRuleChar,
  isRuleCharExclude,
  isRuleEnd,
  isRuleRef,
} from './type-guards.ts';

const KEY_TRANSLATION: Record<RuleType, number> = {
  [RuleType.END]: 0,
  [RuleType.CHAR]: 1,
  [RuleType.CHAR_EXCLUDE]: 2,
  [RuleType.REF]: 3,
};

export const getSerializedRuleKey = (rule: UnresolvedRule): string => {
  if (isRuleEnd(rule)) {
    return `${KEY_TRANSLATION[RuleType.END]}`;
  }

  if (isRuleChar(rule)) {
    return `${KEY_TRANSLATION[RuleType.CHAR]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleCharExclude(rule)) {
    return `${KEY_TRANSLATION[RuleType.CHAR_EXCLUDE]}-${JSON.stringify(rule.value)}`;
  }

  if (isRuleRef(rule)) {
    return `${KEY_TRANSLATION[RuleType.REF]}-${rule.value}`;
  }

  throw new Error(`Unknown rule type: ${JSON.stringify(rule)}`);
};
